### PR TITLE
Make terraform provider generic

### PR DESCRIPTION
**Pull request by eliecharra**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | no
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #213 
| ❓ Documentation  | no

## Description

Remove all related aws stuff from `TerraformProvider` and create an `AWSTerraformProvider`

⚠️ Architecture review please @moadibfr @wbeuil. I splitted commits in small reviewable parts.

### CODE REVIEW

**Review by moadibfr — APPROVED** [review]

**Review by wbeuil — CHANGES_REQUESTED** [review]
Nice rewrite of the architecture to make it more generic to create terraform provider.

My only comment, and I'm really against it, is the renaming of the `region` with `availability_zone`. In AWS, availability zones are really something that is completely different than regions as per the image below.

image

Availability Zones are multiple, isolated locations within each Region. It will be confusing to name it that way.

**Comment by eliecharra** [comment]
> Availability Zones are multiple, isolated locations within each Region. It will be confusing to name it that way.

Agree, I haven't that in mind when looking for a new naming. We should find something generic enough between each provider, maybe we could use `alias` instead ?

**Comment by wbeuil** [comment]
`alias` could do the trick. What do you think @moadibfr ?

**Comment by eliecharra** [comment]
FYI alias came from this : https://www.terraform.io/docs/language/providers/configuration.html#alias-multiple-provider-configurations

File: pkg/remote/aws/provider.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	})
+	if err != nil {
+		return nil, err
+	}
+	p.session = session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}))
+	tfProvider, err := terraform.NewTerraformProvider(installer, terraform.TerraformProviderConfig{
+		DefaultAlias: *p.session.Config.Region,
+		GetProviderConfig: func(availabilityZone string) interface{} {
```
Comment (wbeuil) [inline pkg/remote/aws/provider.go:58]: Should be `alias (string)` here

File: pkg/remote/terraform/provider.go [hunk 2]
```diff
@@ -10,76 +10,43 @@ import (
...
-
-	SkipCredsValidation     bool
-	SkipGetEC2Platforms     bool
-	SkipRegionValidation    bool
-	SkipRequestingAccountId bool
-	SkipMetadataApiCheck    bool
-	S3ForcePathStyle        bool
+type TerraformProviderConfig struct {
+	DefaultAlias      string
+	GetProviderConfig func(alias string) interface{}
```
Comment (wbeuil) [inline pkg/remote/terraform/provider.go:27]: Maybe just above the struct an explanation of what is the alias attribute.

Something like: `Alias refers to the 'meta-region' where the terraform provider will scan all resources`

Comment (eliecharra) [reply pkg/remote/terraform/provider.go:27]: WDYT of 

```
// "alias" in these struct is a way to namespaces gRPC client
// for example, if we need to read S3 bucket from AWS from multiples region, we'll have
// an alias per region, and the alias IS the region itself
// So we can query resources using a specific custom provider configuration
```

Comment (wbeuil) [reply pkg/remote/terraform/provider.go:31]: Nice !

Just quick typo here `"alias" in this ... to namespace ...` `from multiple AWS region, ...`

File: pkg/remote/aws/s3_bucket_inventory_supplier.go [hunk 3]
```diff
@@ -94,7 +94,7 @@ func (s *S3BucketInventorySupplier) listBucketInventoryConfiguration(name, regio
 					Ty: aws.AwsS3BucketInventoryResourceType,
 					ID: id,
 					Attributes: map[string]string{
-						"aws_region": region,
+						"alias: region,
```
Comment (wbeuil) [inline pkg/remote/aws/s3_bucket_inventory_supplier.go:97]: `"alias": region,`

**Review by wbeuil — DISMISSED** [review]

**Pull request merged**